src/goto-programs/goto_program.cpp

```diff
@@ -676,6 +676,50 @@ void goto_programt::instructiont::validate(
   validate_code_full_pick(code, ns, vm);
   validate_expr_full_pick(guard, ns, vm);
 
+  auto evaluates_to_boolean = [](const exprt &e) -> bool {
```

Comment:
The idea was to prevent malformed expression to be passed as guards, e.g. 5+5 with type id `ID_bool`. If we find another use for this check, then moving it to util would be better indeed.

Reply:
Checking whether an expression is malformed or not, especially type checking, seems like a separate (and worthwhile) problem.  It will be useful in many other contexts so I would suggest that it is separately addressed because at the moment I don't see type checking of other exprts as part of this patch set.

Reply:
Ok, not checked here + function moved to `std_expr.cpp`.